### PR TITLE
PD-14397 push content hook

### PR DESCRIPTION
**Pull request by oscarkwan**
### Purpose 🚀
- Create option for sidepanel to push content when opening panel
- Improve animation speeds / smoothness

### Notes ✏️
https://www.loom.com/share/dcdfa262e52d4e94af36dc33ddc4d5e8

### Updates 📦
- [ ] MAJOR (breaking) change to _these packages_
- [x] MINOR (backward compatible) change to _SidePanel_
- [ ] PATCH (bug fix) change to _these packages_

### Storybook 📕
http://storybooks.highbond-s3.com/paprika/PD-14397-push-body-sidepanel

### CODE REVIEW

File: packages/SidePanel/src/SidePanel.js [hunk 1]
```diff
@@ -18,6 +18,9 @@ const propTypes = {
   /** The content for the SidePanel. */
   children: PropTypes.node.isRequired,
 
+  /** Function that provides the container DOM element to be pushed. */
+  getPushContentRef: PropTypes.func,
```
Comment (nahumzs) [inline packages/SidePanel/src/SidePanel.js]: on the popover we are using the following convention when we want to receive an html node object, technically is not a react ref.

  /** Function that provides DOM element to use as target for positioning the popover. */
  getPositioningElement: PropTypes.func,

  /** Function that provides the scrolling DOM element that contains the popover. */
  getScrollContainer: PropTypes.func,

let's adjust the prop name?

Comment (oscarkwan) [reply packages/SidePanel/src/SidePanel.js]: Sure `getPushContentContainer`?

**Comment by nahumzs** [comment]
I have some concern that this pattern might not work correctly on legacy screen where having position absolutes with z-index really hight can appear a the top of the sidepanel or not responsive UI break while resizing it. 

But if the consumer want to pay that toll to have this pattern I 👌 with it.

**Comment by nahumzs** [comment]
Can you add a simple test to have this pattern test? even a screener test with open and close capture might be enough.

File: packages/SidePanel/src/SidePanel.js [hunk 2]
```diff
@@ -121,10 +129,26 @@ function SidePanel(props) {
...
+    pushContentRefStyle.transition = PUSH_REF_TRANSITION_STYLE;
+    pushContentRefStyle.transitionDelay = PUSH_REF_TRANSITION_DELAY_STYLE;
+
+    if (isOpen) {
+      pushContentRefStyle.marginRight = width;
+    } else {
+      pushContentRefStyle.transitionDelay = "0s";
+      pushContentRefStyle.marginRight = "0";
+    }
+  }, [isOpen, getPushContentRef, width]);
```
Comment (AndreyChernykh) [inline packages/SidePanel/src/SidePanel.js]: Do we need a cleanup function?

File: packages/SidePanel/src/SidePanel.js (hunk 1 above)
Comment (nahumzs) [reply packages/SidePanel/src/SidePanel.js]: `getPushContentElement`? I think would be more obvious for the consumer?

**Comment by oscarkwan** [comment]
> I have some concern that this pattern might not work correctly on legacy screen where having position absolutes with z-index really hight can appear a the top of the sidepanel or not responsive UI break while resizing it.
> 
> But if the consumer want to pay that toll to have this pattern I 👌 with it.

I think I am okay with it for now. If there is really absolute positioned elements that affect the layout when pushed, then the consumer should make the judgement call to not push the content

File: packages/SidePanel/stories/examples/Screener.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import React from "react";
+import Screener, { Steps } from "screener-storybook/src/screener";
+import SidePanelPushContentStory from "./SidePanelPushContentStory";
+
+const screenerScript = new Steps()
+  .snapshot("SidePanel - Push content expanded")
```
Comment (nahumzs) [inline packages/SidePanel/stories/examples/Screener.js]: 😮 fancy!!!

**Pull request merged**